server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -650,6 +650,57 @@ public DataStream removeBackingIndex(Index index) {
         );
     }
 
+    /**
+     * Removes the specified failure store index and returns a new {@code DataStream} instance with
+     * the remaining failure store indices.
+     *
+     * @param index the failure store index to remove
+     * @return new {@code DataStream} instance with the remaining failure store indices
+     * @throws IllegalArgumentException if {@code index} is not a failure store index or is the current failure store write index of the
+     * data stream
+     */
+    public DataStream removeFailureStoreIndex(Index index) {
+        int failureIndexPosition = failureIndices.indexOf(index);
+
+        if (failureIndexPosition == -1) {
+            throw new IllegalArgumentException(
+                String.format(Locale.ROOT, "index [%s] is not part of data stream [%s] failure store", index.getName(), name)
+            );
+        }
+
+        // TODO: When failure stores are lazily created, this wont necessarily be required anymore. We can remove the failure store write
+        // index as long as we mark the data stream to lazily rollover the failure store with no conditions on its next write
```

Comment:
I think the lazy rollover is a new enough feature that we haven't considered it before. Failure stores are notable here since we want to make them lazily created, and so a state where the failure store indices list is empty will someday be valid.

I think it might be worth discussing with the wider team about doing something like this for regular data streams, but I don't think it's the most pressing thing. The only thing I can see this helping is cleaning up write indices in data streams that are no longer used, which doesn't solve the fact that after cleaning up the indices, the data stream will still stick around indefinitely until it is deleted.